arch/arc/core/secureshield/arc_secure.S

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2018 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <arch/cpu.h>
+
+.macro clear_scratch_regs
+	mov r1, 0
+	mov r2, 0
+	mov r3, 0
+	mov r4, 0
+	mov r5, 0
+	mov r6, 0
+	mov r7, 0
+	mov r8, 0
+	mov r9, 0
+	mov r10, 0
+	mov r11, 0
+	mov r12, 0
+.endm
+
+.macro clear_callee_regs
+	mov r25, 0
+	mov r24, 0
+	mov r23, 0
+	mov r22, 0
+	mov r21, 0
+	mov r20, 0
+	mov r19, 0
+	mov r18, 0
+	mov r17, 0
+	mov r16, 0
+
+	mov r15, 0
+	mov r14, 0
+	mov r13, 0
+.endm
+
+GTEXT(arc_go_to_normal)
+GTEXT(_arc_do_secure_call)
+GDATA(arc_s_call_table)
+
+
+SECTION_FUNC(TEXT, _arc_do_secure_call)
+	/* r0-r5: arg1-arg6, r6 is call id */
+	/* the call id should be checked */
```

Comment:
not just call id, also other arguments and - depending on the call id/secure system call feature requested - maybe even whether the caller is allowed to call the function.

Reply:
it should be left to the implementation of secure system call handler.